Objects/cown.c

```diff
@@ -0,0 +1,323 @@
+    _Py_MakeImmutable(_PyObject_CAST(Py_TYPE(self)));
+    _Py_notify_regions_in_use();
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown created (%p)\n", self);
+#endif
+
+    sem_init(&self->semaphore, 0, 0);
+    Py_SET_REGION(self, _Py_COWN);
+
+    static char *kwlist[] = {"value", NULL};
+    PyObject *value = NULL;
+
+    // See if we got a value as a keyword argument
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "|O", kwlist, &value)) {
+        return -1;  // Return -1 on failure
+    }
+
+    if (value) {
+        PyCown_set_unchecked(self, value);
+
+    } else {
+        atomic_store(&self->state, Cown_RELEASED);
+        self->value = Py_None;
+    }
+    return 0;
+}
+
+static int PyCown_traverse(PyCownObject *self, visitproc visit, void *arg) {
+    Py_VISIT(self->value);
+    return 0;
+}
+
+#define bail_if_owned(o, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->owning_thread != 0) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S -- %zd", msg, o, o->owning_thread); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_owned(o, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        PyThreadState *tstate = PyThreadState_Get(); \
+        if (o->owning_thread != tstate->thread_id) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_in_state(o, expected_state, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->state != expected_state) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_acquired(o, msg) \
+    bail_unless_owned(o, msg) \
+    bail_unless_in_state(o, Cown_ACQUIRED, msg)
+
+static PyObject *PyCown_acquire(PyCownObject *self, PyObject *args) {
+    int expected = 0;
+
+    while (!atomic_compare_exchange_strong(&self->state, &expected, Cown_ACQUIRED)) {
+        sem_wait(&self->semaphore);
+        expected = 0;
+    }
+
+    // Note: we must hold the GIL at this point -- note for future
+    // threading implementation.
+    PyThreadState *tstate = PyThreadState_Get();
+    self->owning_thread = tstate->thread_id;
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown %p acquired by %zd\n", self, tstate->thread_id);
+#endif
+    Py_RETURN_NONE;
+}
+
+static PyObject *PyCown_release(PyCownObject *self, PyObject *args) {
+    if (self->state == Cown_RELEASED) {
+        bail_if_owned(self, "BUG: Released cown had owning thread: %p");
+        Py_RETURN_NONE;
+    }
+
+    bail_unless_owned(self, "Thread attempted to release a cown it did not own");
+
+    atomic_store(&self->state, Cown_RELEASED);
+    self->owning_thread = 0;
```

Comment:
This feels safer the other way around.
```suggestion
    self->owning_thread = 0;
    atomic_store(&self->state, Cown_RELEASED);
```

Reply:
Indeed! 😅